abarnash/moose-stack
# Create games

Allow a user to create and join a game
Pull out queries and mutations to their own file as they can only be defined once
Pull out common queries and responses 

### Testing plan
1. Start ya server
```bash
cd containers/moose
npm start
```
2. Navigate to the graphql playground at http://localhost:4000/
3. Create a user by pasting the following into the left side of the playground, clicking the play button, and selecting AddUser
```
mutation AddUser {
  newUser(username: "game-tester", name: "Game Tester", email: "game-tester@gmail.com") {
    success
    message
    user {
      username
    }
  }
}
```
4. Confirm the following output
```json
{
  "data": {
    "newUser": {
      "success": true,
      "message": "New user with username \"game-tester\" created!",
      "user": {
        "username": "game-tester"
      }
    }
  }
}
```
5. Log in as the user by running the following mutation and copying the returned token
```
mutation Login {
  login(username: "game-tester") {
    token
  }
}
```
```json
{
  "data": {
    "login": {
      "token": "Z2FtZS10ZXN0ZXI="
    }
  }
}
```
6. Add the token to the HTTP HEADERS tab at the bottom of the playground
```json
{
  "authorization": "Z2FtZS10ZXN0ZXI="
}
```
7. Confirm your user is logged in
```
query GetLoggedInUser {
  loggedInUser {
    id
    username
    name
    email
    currentGameUrl
  }
}
```
```json
{
  "data": {
    "loggedInUser": {
      "id": "33e282a0-ce79-430e-8a05-b7303a91f717",
      "username": "game-tester",
      "name": "Game Tester",
      "email": "game-tester@gmail.com",
      "currentGameUrl": null
    }
  }
}
```
8. Create a game!
```
mutation AddGame {
  newGame {
    success
    message
    game {
      url
    }
  }
}
```
```json
{
  "data": {
    "newGame": {
      "success": true,
      "message": "New game with url \"of-idea\" created!",
      "game": {
        "url": "of-idea"
      }
    }
  }
}
```
7. Join the game!
```
mutation JoinGame {
  joinGame(url: "of-idea") {
    message
    success
    game {
      url
    }
  }
}
```
```json
{
  "data": {
    "joinGame": {
      "message": "User game-tester has joined game of-idea",
      "success": true,
      "game": {
        "url": "of-idea"
      }
    }
  }
}
```
8. Confirm the user's current game was updated by running GetLoggedInUser again and confirming `currentGameUrl` matches the 
```json
{
  "data": {
    "loggedInUser": {
      "id": "33e282a0-ce79-430e-8a05-b7303a91f717",
      "username": "game-tester",
      "name": "Game Tester",
      "email": "game-tester@gmail.com",
      "currentGameUrl": "of-idea"
    }
  }
}
```
9. Try to join a non-existent game
```
mutation JoinGame {
  joinGame(url: "not-a-game") {
    message
    success
    game {
      url
    }
  }
}
```
```json
{
  "data": {
    "joinGame": {
      "message": "Cannot find game with url \"not-a-game\"",
      "success": false,
      "game": null
    }
  }
}
```
10. Try to join a game when not logged in by removing the http header
```json
{
  "data": {
    "joinGame": {
      "message": "User must be logged in to perform this action",
      "success": false,
      "game": null
    }
  }
}
```
